src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionLiteral.cs

```diff
@@ -12,44 +14,81 @@ namespace Microsoft.CodeAnalysis.CSharp
+                case CollectionLiteralTypeKind.Array:
+                case CollectionLiteralTypeKind.Span:
+                case CollectionLiteralTypeKind.ReadOnlySpan:
+                    Debug.Assert(elementType is { });
+                    return VisitArrayOrSpanCollectionLiteralExpression(node, elementType);
+                case CollectionLiteralTypeKind.ListInterface:
+                    return VisitListInterfaceCollectionLiteralExpression(node);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(collectionTypeKind);
+            }
+        }
+
+        private BoundExpression VisitArrayOrSpanCollectionLiteralExpression(BoundCollectionLiteralExpression node, TypeSymbol elementType)
         {
             Debug.Assert(!_inExpressionLambda);
             Debug.Assert(node.Type is { });
 
             var syntax = node.Syntax;
             var collectionType = node.Type;
-            var spanConstructor = node.SpanConstructor;
+            MethodSymbol? spanConstructor = null;
 
             var arrayType = collectionType as ArrayTypeSymbol;
             if (arrayType is null)
             {
-                Debug.Assert(spanConstructor is { });
-                Debug.Assert(TypeSymbol.Equals(collectionType, spanConstructor.ContainingType, TypeCompareKind.AllIgnoreOptions));
-
-                var elementType = ((NamedTypeSymbol)collectionType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
-                arrayType = ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, elementType);
+                Debug.Assert(collectionType.Name is "Span" or "ReadOnlySpan");
+                // We're constructing a Span<T> or ReadOnlySpan<T> rather than T[].
+                var spanType = (NamedTypeSymbol)collectionType;
+                arrayType = ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, spanType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0]);
+                spanConstructor = ((MethodSymbol)_compilation.GetWellKnownTypeMember(
+                    collectionType.Name == "Span" ? WellKnownMember.System_Span_T__ctor_Array : WellKnownMember.System_ReadOnlySpan_T__ctor_Array)!).AsMember(spanType);
             }
 
-            var initializers = node.Initializers;
-            int arrayLength = initializers.Length;
-            // PROTOTYPE: Should [] be emitted as Array.Empty<T>()?
-            var initialization = (arrayLength == 0)
-                ? null
-                : new BoundArrayInitialization(
-                    syntax,
-                    isInferred: false,
-                    initializers.SelectAsArray(e => VisitExpression(e)));
-            var array = new BoundArrayCreation(
-                syntax,
-                ImmutableArray.Create<BoundExpression>(
-                    new BoundLiteral(
+            var elements = node.Elements;
+            BoundExpression array;
+
+            if (elements.Any(i => i is BoundCollectionLiteralSpreadElement))
+            {
```

Comment:
I wonder if perhaps the `BoundCollectionLiteral` should directly store its length as a non-negative int if it is known, or -1 if the length is unknown.